Support pickling of some objects

GodTamIt opened the pull request:
Fixes #96, supporting serialization via `pythonize`.

The (personal) use-case here is to be able to send types across process boundaries (i.e. large batches of `Document`s across to a process pool for some kind of processing).

cjrh commented:
Hi @GodTamIt 

Thank you for this! A few general comments:

The maintainers here have very little bandwidth, so larger PRs can take a lot of time to get reviewed. Don't get discouraged if it takes a while, this is normal. Generally speaking, smaller PRs travel exponentially faster than large PRs, in many repos not only this one.

The PR includes extra unnecessary changes like renaming `Value` to `TvValue`. I prefer large renames and similar refactors like this to be in separate PRs. (and if this particular rename was in its own PR, I would reject the PR as unnecessary).

I feel like the motivation for supporting pickle feels not yet strong enough to support a change of this size, and the taking on of new dependencies (in my opinion). A use-case would help to understand the motivation better.  Is it to be able to move Fruit between processes/machines? Or a distributed system or similar? If we had our documentation already in place I'd ask you to add docs about the pickle feature which would help to explain more, but the docs aren't quite ready yet (#94).

This PR adds new dependencies. This will add to the time for review because I will go look at each of the dependencies to look for reasons to reject them. I have a high bar for taking on new dependencies. They are easy to add and very hard to remove. For instance, flexbuffers is a heavy C++ dependency we take on, just to support pickle. I would like us to try harder to make bincode work, or pythonize as suggested in #96? Or perhaps [serde_pickle](https://docs.rs/serde-pickle/latest/serde_pickle/), etc.

I see you added the time crate, but seeing as we already have chrono, did you look at [chrono::serde](https://docs.rs/chrono/latest/chrono/serde/)?

GodTamIt commented:
> Hi @GodTamIt
> 
> Thank you for this! A few general comments:
> 
> The maintainers here have very little bandwidth, so larger PRs can take a lot of time to get reviewed. Don't get discouraged if it takes a while, this is normal. Generally speaking, smaller PRs travel exponentially faster than large PRs, in many repos not only this one.
>
> The PR includes extra unnecessary changes like renaming `Value` to `TvValue`. I prefer large renames and similar refactors like this to be in separate PRs. (and if this particular rename was in its own PR, I would reject the PR as unnecessary).

Ack. I appreciate the feedback here, and I've removed these changes. In the future, I will try to limit the change to only the necessary changes.

> I feel like the motivation for supporting pickle feels not yet strong enough to support a change of this size, and the taking on of new dependencies (in my opinion). A use-case would help to understand the motivation better. Is it to be able to move Fruit between processes/machines? Or a distributed system or similar? If we had our documentation already in place I'd ask you to add docs about the pickle feature which would help to explain more, but the docs aren't quite ready yet (#94).

The primary motivation here is to be able to send these types (`Document`, `SearchResult`, etc.) across a process boundary. For example, `concurrent.futures.ProcessPoolExecutor` to operate on huge batches of `Document`s.

> This PR adds new dependencies. This will add to the time for review because I will go look at each of the dependencies to look for reasons to reject them. I have a high bar for taking on new dependencies. They are easy to add and very hard to remove. For instance, flexbuffers is a heavy C++ dependency we take on, just to support pickle. I would like us to try harder to make bincode work, or pythonize as suggested in #96? Or perhaps [serde_pickle](https://docs.rs/serde-pickle/latest/serde_pickle/), etc.

Ah, I didn't see the `pythonize`suggestion until after posting this PR. I've elected for that, as it seems to be far lighter as a dependency and something that works.

> I see you added the time crate, but seeing as we already have chrono, did you look at [chrono::serde](https://docs.rs/chrono/latest/chrono/serde/)?

`tantivy` opts for the `time` crate internally instead. However, since `tantivy` likes to store it as a nanosecond timestamp, I've changed the code to just use an `i64`.

GodTamIt commented on Cargo.toml:
```diff
@@ -18,8 +18,10 @@ chrono = "0.4.23"
 [dependencies.pyo3]
-version = "0.18.0"
-features = ["chrono", "extension-module"]
+version = "0.19.1"
```
This change needs to wait for #98 to land.

cjrh replied:
It's in 👍🏼

GodTamIt commented:
This change was stacked on top of #99 (for equality support) and will need #98 to land before it can go in.

adamreichold commented on Cargo.toml:
```diff
@@ -18,6 +18,8 @@ chrono = "0.4.23"
 itertools = "0.10.5"
 futures = "0.3.26"
+pythonize = "0.19.0"
+serde = "~1.0"
```
Why a non-default version specification for the requirement on `serde`?

adamreichold replied:
Sorry, I was unspecific. Why do you use a tilde requirement `~1.0` meaning `>=1.0, <1.1` instead of the default type `1.0` meaning `^1.0` meaning `>=1.0, <2.0`?

GodTamIt replied:
I wanted to be more conservative/more explicit with the permissible versions, since unexpected changes in `serde` could make previously pickled types might not be compatible. But `1.0` (aka `^1.0`) should be fine, since breaking changes aren't likely with minor version changes.

adamreichold commented on src/document.rs:
```diff
@@ -106,6 +111,98 @@ fn value_to_string(value: &Value) -> String {
+    i64::deserialize(deserializer).map(tv::DateTime::from_timestamp_nanos)
+}
+
+/// An equivalent type to [`tantivy::schema::Value`] but uses tagging in its serialization to
```
I am not sure I understand why this requires a different serialization schema compared to the upstream `Value` type.

adamreichold commented on src/document.rs:
```diff
@@ -282,6 +379,67 @@ impl fmt::Debug for Document {
+    {
+        struct MapVisitor;
+
+        impl<'de> Visitor<'de> for MapVisitor {
```
Wouldn't it be easier to just deserialize `BTreeMap<String, Vec<SerdeValue>>` and then convert this to `BTreeMap<String, Vec<Value>>`?

adamreichold replied:
(I think the re-allocation could even be avoid if the conversion uses `.into_iter().map(..).collect()` as the types should layout compatible.)

GodTamIt replied:
These are great suggestions, done.

GodTamIt commented on src/document.rs:
```diff
@@ -282,6 +379,67 @@ impl fmt::Debug for Document {
+            serializer.serialize_map(Some(self.field_values.len()))?;
+        for (k, v) in &self.field_values {
+            let ser_v: Vec<_> =
+                v.iter().cloned().map(SerdeValue::from).collect();
```
These are great suggestions, thanks!

adamreichold commented on src/schema.rs:
```diff
@@ -25,4 +27,24 @@ impl Schema {
     }
+
+    #[staticmethod]
+    fn _internal_from_pythonized(serialized: &PyAny) -> PyResult<Self> {
```
Agreed. The social convention should suffice. Thanks for looking into it!

adamreichold commented on src/document.rs:
```diff
@@ -282,6 +428,42 @@ impl fmt::Debug for Document {
+    where
+        D: Deserializer<'de>,
+    {
+        HashMap::<String, Vec<SerdeValue>>::deserialize(deserializer).map(
```
I would recommend sticking to `BTreeMap` as the conversion from `HashMap` to `BTreeMap` will definitely need to re-allocate whereas `BTreeMap` could potentially avoid it.

GodTamIt replied:
Oops, good catch, done.

GodTamIt commented:
Thanks for the reviews! This can land whenever
